src/Symfony/Component/DependencyInjection/Tests/Compiler/DecoratorServicePassTest.php

```diff
@@ -115,6 +115,36 @@ public function testProcessWithPriority()
         $this->assertNull($quxDefinition->getDecoratedService());
     }
 
+    public function testProcessWithDeprecation()
+    {
+        $container = new ContainerBuilder();
+        $fooDefinition = $container
+            ->register('deprecated_foo')
+            ->setPublic(false)
+        ;
+        $barDefinition = $container
+            ->register('bar')
+            ->setPublic(true)
+            ->setDeprecated(true)
+            ->setDecoratedService('deprecated_foo')
+        ;
+
+        $wasTriggered = false;
+
+        $handler = set_error_handler(function ($errno, $errstr) use (&$wasTriggered) {
+            $this->assertSame(E_USER_DEPRECATED, $errno);
+            $this->assertSame('The service deprecated_foo is deprecated. You should stop using it, as it will soon be removed.', $errstr);
+
+            $wasTriggered = true;
+        });
+
+        $this->process($container);
+
+        set_error_handler($handler);
```

Comment:
you should use ``restore_error_handler`` to restore the previous one, instead of adding a new one in the stack